aya/src/programs/links.rs

```diff
@@ -394,6 +397,174 @@ pub enum LinkError {
+
+    /// Ensure the link is created before the specified aya-owned link for a given attachment point.
+    pub fn before_link<L: MultiProgLink>(link: &L) -> Result<Self, LinkError> {
+        Ok(Self {
+            link_ref: LinkRef::Fd(link.fd()?.as_raw_fd()),
+            flags: MprogFlags::BEFORE | MprogFlags::LINK,
+            expected_revision: None,
+        })
+    }
+
+    /// Ensure the link is created after the specified aya-owned link for a given attachment point.
+    pub fn after_link<L: MultiProgLink>(link: &L) -> Result<Self, LinkError> {
+        Ok(Self {
+            link_ref: LinkRef::Fd(link.fd()?.as_raw_fd()),
+            flags: MprogFlags::AFTER | MprogFlags::LINK,
+            expected_revision: None,
+        })
+    }
+
+    /// Ensure the link is created before a link specified by its kernel id for a given attachment point.
+    pub fn before_link_id(id: LinkId) -> Result<Self, LinkError> {
+        Ok(Self {
+            link_ref: LinkRef::Id(id.0),
+            flags: MprogFlags::BEFORE | MprogFlags::LINK | MprogFlags::ID,
+            expected_revision: None,
+        })
+    }
+
+    /// Ensure the link is created after a link specified by its kernel id for a given attachment point.
+    pub fn after_link_id(id: LinkId) -> Result<Self, LinkError> {
+        Ok(Self {
+            link_ref: LinkRef::Id(id.0),
+            flags: MprogFlags::AFTER | MprogFlags::LINK | MprogFlags::ID,
+            expected_revision: None,
+        })
+    }
+
+    /// Ensure the link is created before the specified aya-owned program for a given attachment point.
+    pub fn before_program<P: MultiProgProgram>(program: &P) -> Result<Self, ProgramError> {
+        Ok(Self {
+            link_ref: LinkRef::Fd(program.fd()?.as_raw_fd()),
+            flags: MprogFlags::BEFORE,
+            expected_revision: None,
+        })
+    }
+
+    /// Ensure the link is created after the specified aya-owned program for a given attachment point.
+    pub fn after_program<P: MultiProgProgram>(program: &P) -> Result<Self, ProgramError> {
+        Ok(Self {
+            link_ref: LinkRef::Fd(program.fd()?.as_raw_fd()),
+            flags: MprogFlags::AFTER,
+            expected_revision: None,
+        })
+    }
+
+    /// Ensure the link is created before a program specified by its kernel id for a given attachment point.
+    pub fn before_program_id(id: ProgramId) -> Self {
+        Self {
+            link_ref: LinkRef::Id(id.0),
+            flags: MprogFlags::BEFORE | MprogFlags::ID,
+            expected_revision: None,
+        }
+    }
+
+    /// Ensure the link is created after a program specified by its kernel id for a given attachment point.
+    pub fn after_program_id(id: ProgramId) -> Self {
+        Self {
+            link_ref: LinkRef::Id(id.0),
+            flags: MprogFlags::AFTER | MprogFlags::ID,
+            expected_revision: None,
+        }
+    }
+
+    /// set the expected revision for the link, the revision changes
```

Comment:
The revision is maintained per attach point (interface + direction) and is incremented each time you attach or detach a program to that point. expected_revision allows you to ensure that nothing has changed between the time you checked the state of the given attach point and when your tcx program is attached.  In particular, if the current revision is different from the provided expected_revision, the attach operation will fail.

So, you are correct that there needs to be a way to get the current revision in order to use it effectively.  The current state of an attach point (including revision) can be retrieved using the BPF_PROG_QUERY syscall.  This is an optional attribute. I’ve tested the behavior of expected_revision with the current pr and it works correctly.  However, since it’s optional, and we’re not using it in the initial bpfman implementation, I propose that I add support for getting the state via the BPF_PROG_QUERY syscall in a follow-on pr.